infrastructure/cli/commands/s3/put.ts

```diff
@@ -0,0 +1,46 @@
+import * as fs from 'fs/promises'
+import * as path from 'path'
+
+import {Client} from './index.js'
+import {makeCommand} from '../../lib/command.js'
+import {collect} from '../../lib/misc.js'
+import {EnvironmentOptions} from '../../lib/options.js'
+
+export const {command, description, builder} = makeCommand(
+  import.meta.url,
+  'Upload objects to S3',
+  {
+    requiredPositionals: [
+      {
+        name: 'key',
+        description: 'Object key',
+      },
+    ],
+    optionalPositionals: [
+      {
+        name: 'open-from',
+        description: 'File to upload (defaults to stdin)',
+      },
+    ],
+  }
+)
+
+interface Options extends EnvironmentOptions {
+  key: string
+  openFrom?: string
+}
+
+export async function handler({environment, key, openFrom}: Options): Promise<void> {
+  const client = await Client.load(environment)
+  // eslint-disable-next-line init-declarations
+  let content
+  if (openFrom) {
+    const openFromPath = path.resolve(openFrom)
+    const readHandle = await fs.open(openFromPath, 'r')
+    content = await readHandle.readFile()
+  } else {
+    const chunks: any[] = await collect(process.stdin)
+    content = Buffer.concat(chunks)
+  }
```

Comment:
I think `putObject()`'s `data` can also be a `stream.Readable`, and `process.stdout` should be a `stream.Duplex` (which is a fancy way of saying it's a `stream.Readable` and a `stream.Writable`), anyway...

This means that potentially (I haven't tested sorry) you could write something like:

```javascript
let readable

if (openFrom) {
  readable = fs.createReadStream(openFromPath)
} else {
  readable = process.stdin
}

await client.putObject(key, readable)
```

You may even squeeze a ternary operator in there (but maybe it's too terse?):

```javascript
let readable = openFrom ? fs.createReadStream(openFromPath) : process.stdin

await client.putObject(key, readable)
```

Again, not tested but in theory it _should_ work 🤷

Reply:
yeah i agree that your version is better because it doesn’t buffer the whole input content before passing it to S3 client